cpp/src/neighbors/ann_mg/ann_mg.cuh

```diff
@@ -0,0 +1,750 @@
+      auto& ann_if = ann_interfaces_[rank];
+      ann_if.build(dev_res, index_params, partition);
+      resource::sync_stream(dev_res);
+    }
+#pragma omp barrier
+  }
+}
+
+template <typename AnnIndexType, typename T, typename IdxT>
+void ann_mg_index<AnnIndexType, T, IdxT>::extend(
+  const cuvs::neighbors::mg::nccl_clique& clique,
+  raft::host_matrix_view<const T, IdxT, row_major> new_vectors,
+  std::optional<raft::host_vector_view<const IdxT, IdxT>> new_indices)
+{
+  IdxT n_rows = new_vectors.extent(0);
+  if (mode_ == REPLICATED) {
+    RAFT_LOG_INFO("REPLICATED EXTEND: %d*%drows", num_ranks_, n_rows);
+
+#pragma omp parallel for
+    for (int rank = 0; rank < num_ranks_; rank++) {
+      int dev_id                            = clique.device_ids_[rank];
+      const raft::device_resources& dev_res = clique.device_resources_[rank];
+      RAFT_CUDA_TRY(cudaSetDevice(dev_id));
+      auto& ann_if = ann_interfaces_[rank];
+      ann_if.extend(dev_res, new_vectors, new_indices);
+      resource::sync_stream(dev_res);
+    }
+#pragma omp barrier
+  } else if (mode_ == SHARDED) {
+    IdxT n_cols           = new_vectors.extent(1);
+    IdxT n_rows_per_shard = raft::ceildiv(n_rows, (IdxT)num_ranks_);
+
+    RAFT_LOG_INFO("SHARDED EXTEND: %d*%drows", num_ranks_, n_rows_per_shard);
+
+#pragma omp parallel for
+    for (int rank = 0; rank < num_ranks_; rank++) {
+      int dev_id                            = clique.device_ids_[rank];
+      const raft::device_resources& dev_res = clique.device_resources_[rank];
+      RAFT_CUDA_TRY(cudaSetDevice(dev_id));
+      IdxT offset                  = rank * n_rows_per_shard;
+      IdxT n_rows_of_current_shard = std::min(n_rows_per_shard, n_rows - offset);
+      const T* new_vectors_ptr     = new_vectors.data_handle() + (offset * n_cols);
+      auto new_vectors_part        = raft::make_host_matrix_view<const T, IdxT, row_major>(
+        new_vectors_ptr, n_rows_of_current_shard, n_cols);
+
+      std::optional<raft::host_vector_view<const IdxT, IdxT>> new_indices_part = std::nullopt;
+      if (new_indices.has_value()) {
+        const IdxT* new_indices_ptr = new_indices.value().data_handle() + offset;
+        new_indices_part =
+          raft::make_host_vector_view<const IdxT, IdxT>(new_indices_ptr, n_rows_of_current_shard);
+      }
+      auto& ann_if = ann_interfaces_[rank];
+      ann_if.extend(dev_res, new_vectors_part, new_indices_part);
+      resource::sync_stream(dev_res);
+    }
+#pragma omp barrier
+  }
+}
+
+template <typename AnnIndexType, typename T, typename IdxT>
+void ann_mg_index<AnnIndexType, T, IdxT>::search(
+  const cuvs::neighbors::mg::nccl_clique& clique,
+  const cuvs::neighbors::search_params* search_params,
+  raft::host_matrix_view<const T, IdxT, row_major> queries,
+  raft::host_matrix_view<IdxT, IdxT, row_major> neighbors,
+  raft::host_matrix_view<float, IdxT, row_major> distances,
+  IdxT n_rows_per_batch) const
```

Comment:
Do we really need this argument? Internally our ANN methods do batching if necessary, why do we need another level of batching here?

Reply:
This is to avoid getting an OOM error on the GPU running the merges. This value is set very high as a default, but if there's any issue, it will be as simple as lowering this value. In the end, when scaling on very large setups (multi nodes), we might opt for a tree merge which is less hungry for memory, but requires more steps. But this would be for a follow-up PR.

Reply:
Ok, thanks for the explanation!